File: workers/java/io/temporal/omes/ClientActionExecutor.java
package io.temporal.omes;

import io.temporal.api.enums.v1.WorkflowIdConflictPolicy;
import io.temporal.client.UpdateOptions;
import io.temporal.client.WorkflowClient;
import io.temporal.client.WorkflowOptions;
import io.temporal.client.WorkflowStub;
import io.temporal.failure.ApplicationFailure;

public class ClientActionExecutor {

  private final WorkflowClient client;
  private final String workflowId;
  private final String workflowType = "kitchenSink";
  private final Object workflowInput = null;
  private final String taskQueue;
  private final boolean errOnUnimplemented;

  public ClientActionExecutor(WorkflowClient client, String workflowId, String taskQueue) {
    this(client, workflowId, taskQueue, false);
  }

  public ClientActionExecutor(
      WorkflowClient client, String workflowId, String taskQueue, boolean errOnUnimplemented) {
    this.client = client;
    this.workflowId = workflowId;
    this.taskQueue = taskQueue;
    this.errOnUnimplemented = errOnUnimplemented;
  }

  public void executeClientSequence(KitchenSink.ClientSequence clientSeq) {
    for (KitchenSink.ClientActionSet actionSet : clientSeq.getActionSetsList()) {
      executeClientActionSet(actionSet);
    }
  }

  private void executeClientActionSet(KitchenSink.ClientActionSet actionSet) {
    if (actionSet.getConcurrent()) {
      if (errOnUnimplemented) {
        throw ApplicationFailure.newNonRetryableFailure(
            "concurrent client actions are not supported", "UnsupportedOperation");
      }
      // Skip concurrent actions when not erroring on unimplemented
      System.out.println("Skipping concurrent client actions (not implemented)");
      return;
    }

    for (KitchenSink.ClientAction action : actionSet.getActionsList()) {
      executeClientAction(action);
    }
  }

  private void executeClientAction(KitchenSink.ClientAction action) {
    if (action.hasDoSignal()) {
      executeSignalAction(action.getDoSignal());
    } else if (action.hasDoUpdate()) {
      executeUpdateAction(action.getDoUpdate());
    } else if (action.hasDoQuery()) {
      executeQueryAction(action.getDoQuery());
    } else if (action.hasNestedActions()) {
      executeClientActionSet(action.getNestedActions());
    } else {
      throw new IllegalArgumentException("Client action must have a recognized variant");
    }
  }

  private void executeSignalAction(KitchenSink.DoSignal signal) {
    String signalName;
    Object signalArgs;

    if (signal.hasDoSignalActions()) {
      signalName = "do_actions_signal";
      signalArgs = signal.getDoSignalActions();
    } else if (signal.hasCustom()) {
      signalName = signal.getCustom().getName();
      signalArgs = signal.getCustom().getArgsList().toArray();
    } else {
      throw new IllegalArgumentException("DoSignal must have a recognizable variant");
    }

    if (signal.getWithStart()) {
      WorkflowOptions options =
          WorkflowOptions.newBuilder()
              .setWorkflowId(workflowId)
              .setTaskQueue(taskQueue)
              .setWorkflowIdConflictPolicy(
                  WorkflowIdConflictPolicy.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING)
              .build();
      WorkflowStub stub = client.newUntypedWorkflowStub(workflowType, options);
      stub.signalWithStart(signalName, new Object[] {signalArgs}, new Object[] {workflowInput});
    } else {
      WorkflowStub stub = client.newUntypedWorkflowStub(workflowId);
      stub.signal(signalName, signalArgs);
    }
  }

  private void executeUpdateAction(KitchenSink.DoUpdate update) {
    String updateName;
    Object updateArgs;

    if (update.hasDoActions()) {
      updateName = "do_actions_update";
      updateArgs = update.getDoActions();
    } else if (update.hasCustom()) {
      updateName = update.getCustom().getName();
      updateArgs = update.getCustom().getArgsList().toArray();
    } else {
      throw new IllegalArgumentException("DoUpdate must have a recognizable variant");
    }

    try {
      if (update.getWithStart()) {
        WorkflowOptions workflowOptions =
            WorkflowOptions.newBuilder()
                .setWorkflowId(workflowId)
                .setTaskQueue(taskQueue)
                .setWorkflowIdConflictPolicy(
                    WorkflowIdConflictPolicy.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING)
                .build();
        WorkflowStub stub = client.newUntypedWorkflowStub(workflowType, workflowOptions);

        UpdateOptions<KitchenSink.WorkflowState> updateOptions =
            UpdateOptions.newBuilder(KitchenSink.WorkflowState.class)
                .setUpdateName(updateName)
                .build();

        stub.executeUpdateWithStart(
            updateOptions, new Object[] {updateArgs}, new Object[] {workflowInput});
      } else {
        WorkflowStub stub = client.newUntypedWorkflowStub(workflowId);
        stub.update(updateName, KitchenSink.WorkflowState.class, updateArgs);
      }
    } catch (Exception e) {
      if (!update.getFailureExpected()) {
        throw e;
      }
      // If failure was expected, swallow the exception
    }
  }

  private void executeQueryAction(KitchenSink.DoQuery query) {
    try {
      WorkflowStub stub = client.newUntypedWorkflowStub(workflowId);

      if (query.hasReportState()) {
        stub.query("report_state", KitchenSink.WorkflowState.class, (Object) null);
      } else if (query.hasCustom()) {
        String queryName = query.getCustom().getName();
        Object[] queryArgs = query.getCustom().getArgsList().toArray();
        stub.query(queryName, Object.class, queryArgs);
      } else {
        throw new IllegalArgumentException("DoQuery must have a recognizable variant");
      }
    } catch (Exception e) {
      if (!query.getFailureExpected()) {
        throw e;
      }
      // If failure was expected, swallow the exception
    }
  }
}

File: workers/typescript/src/client-action-executor.ts
import { Client, WithStartWorkflowOperation } from '@temporalio/client';
import { ApplicationFailure } from '@temporalio/common';
import { WorkflowIdConflictPolicy } from '@temporalio/client';
import { temporal } from './protos/root';
import IClientSequence = temporal.omes.kitchen_sink.IClientSequence;
import IClientActionSet = temporal.omes.kitchen_sink.IClientActionSet;
import IClientAction = temporal.omes.kitchen_sink.IClientAction;
import IDoSignal = temporal.omes.kitchen_sink.IDoSignal;
import IDoUpdate = temporal.omes.kitchen_sink.IDoUpdate;
import IDoQuery = temporal.omes.kitchen_sink.IDoQuery;

export class ClientActionExecutor {
  private client: Client;
  private workflowId = '';
  private runId = '';
  private workflowType = 'kitchenSink';
  private workflowInput: any = null;
  private taskQueue;
  private errOnUnimplemented: boolean;

  constructor(client: Client, workflowId: string, taskQueue: string, errOnUnimplemented: boolean = false) {
    this.client = client;
    this.workflowId = workflowId;
    this.taskQueue = taskQueue;
    this.errOnUnimplemented = errOnUnimplemented;
  }

  async executeClientSequence(clientSeq?: IClientSequence | null): Promise<void> {
    for (const actionSet of clientSeq?.actionSets || []) {
      await this.executeClientActionSet(actionSet);
    }
  }

  private async executeClientActionSet(actionSet: IClientActionSet): Promise<void> {
    if (actionSet.concurrent) {
      if (this.errOnUnimplemented) {
        throw ApplicationFailure.nonRetryable('concurrent client actions are not supported');
      }
      // Skip concurrent actions when not erroring on unimplemented
      return;
    }

    for (const action of actionSet.actions || []) {
      await this.executeClientAction(action);
    }
  }

  private async executeClientAction(action: IClientAction): Promise<void> {
    if (action.doSignal) {
      await this.executeSignalAction(action.doSignal);
    } else if (action.doUpdate) {
      await this.executeUpdateAction(action.doUpdate);
    } else if (action.doQuery) {
      await this.executeQueryAction(action.doQuery);
    } else if (action.nestedActions) {
      await this.executeClientActionSet(action.nestedActions);
    } else {
      throw new Error('Client action must have a recognized variant');
    }
  }

  private async executeSignalAction(signal: IDoSignal): Promise<void> {
    let signalName: string;
    let signalArgs: any;

    if (signal.doSignalActions) {
      signalName = 'do_actions_signal';
      signalArgs = [signal.doSignalActions];
    } else if (signal.custom) {
      signalName = signal.custom.name || '';
      signalArgs = signal.custom.args || [];
    } else {
      throw new Error('DoSignal must have a recognizable variant');
    }

    try {
      if (signal.withStart) {
        const handle = await this.client.workflow.signalWithStart(this.workflowType, {
          workflowId: this.workflowId,
          taskQueue: this.taskQueue,
          args: [this.workflowInput],
          signal: signalName,
          signalArgs,
          workflowIdConflictPolicy: WorkflowIdConflictPolicy.USE_EXISTING,
        });
        this.workflowId = handle.workflowId;
        this.runId = handle.signaledRunId;
      } else {
        const handle = this.client.workflow.getHandle(this.workflowId);
        await handle.signal(signalName, ...signalArgs);
      }
    } catch (error) {
      console.error(`Signal execution failed for ${signalName}:`, error);
      throw error;
    }
  }

  private async executeUpdateAction(update: IDoUpdate): Promise<void> {
    let updateName: string;
    let updateArgs: any;

    if (update.doActions) {
      updateName = 'do_actions_update';
      updateArgs = update.doActions;
    } else if (update.custom) {
      updateName = update.custom.name || '';
      updateArgs = update.custom.args || [];
    } else {
      throw new Error('DoUpdate must have a recognizable variant');
    }

    try {
      if (update.withStart) {
        const startWorkflowOperation = new WithStartWorkflowOperation(this.workflowType, {
          workflowId: this.workflowId,
          taskQueue: this.taskQueue,
          args: [this.workflowInput],
          workflowIdConflictPolicy: WorkflowIdConflictPolicy.USE_EXISTING,
        });
        await this.client.workflow.executeUpdateWithStart(updateName, {
          args: [updateArgs],
          startWorkflowOperation,
        });
      } else {
        const handle = this.client.workflow.getHandle(this.workflowId);
        await handle.executeUpdate(updateName, { args: [updateArgs] });
      }
    } catch (error) {
      console.error(`Update execution failed for ${updateName}:`, error);
      if (!update.failureExpected) {
        throw error;
      }
    }
  }

  private async executeQueryAction(query: IDoQuery): Promise<void> {
    try {
      if (query.reportState) {
        const handle = this.client.workflow.getHandle(this.workflowId);
        await handle.query('report_state', null);
      } else if (query.custom) {
        const handle = this.client.workflow.getHandle(this.workflowId);
        const queryArgs = query.custom.args || [];
        await handle.query(query.custom.name || '', ...queryArgs);
      } else {
        throw new Error('DoQuery must have a recognizable variant');
      }
    } catch (error: unknown) {
      if (!query.failureExpected) {
        throw error;
      }
    }
  }
}
